crates/engine/tree/src/tree/mod.rs

```diff
@@ -2680,10 +2681,42 @@ where
     ) -> Result<TrieInput, ParallelStateRootError> {
         let mut input = TrieInput::default();
 
-        if let Some((historical, blocks)) = self.state.tree_state.blocks_by_hash(parent_hash) {
+        let provider = consistent_view.provider_ro()?;
+        let best_block_number = provider.best_block_number()?;
+
+        let in_memory_blocks = self
+            .state
+            .tree_state
+            .blocks_by_hash(parent_hash)
+            .map(|(_, mut blocks)| {
+                // Iterate over the blocks from oldest to newest.
+                while let Some(block) = blocks.last() {
+                    let recovered_block = block.recovered_block();
+                    // Remove those blocks that lower than or equal to the highest database block
+                    // and have an associated database block with a matching hash.
+                    if recovered_block.number() <= best_block_number &&
+                        Some(recovered_block.hash()) ==
+                            provider.block_hash(recovered_block.number())?
```

Comment:
have haven't checked this before why are we doing this now?

Reply:
does that make sense? https://github.com/paradigmxyz/reth/blob/d8c6d99be572261e6ddfb7fc620f897bc47958c1/crates/engine/tree/src/tree/mod.rs#L2698-L2702

Reply:
all of this is very confusing, 
we at least need more docs on what all of this even is and why we're doing this

it's very hard for me to understand why we're even doing this